opt: introduce invisible columns to opt catalog

RaduBerinde opened the pull request:
There is a need to introduce a new column state during schema changes:
a state in which a column has been backfilled and is readable for
internal operations (like updating an index on that column during an
UPDATE), but is still not public.

Adding a new ColumnKind would complicate a lot of the optimizer code
and add more combinations that need testing. To avoid this, we note
that there is effectively no difference between such a column and an
ordinary column, other than it not being visible to queries.

This change extends the `IsHidden()` column property to a
`ColumnVisibility` which has three values (Visible, Hidden,
Invisible). The new column state will be presented as an Ordinary,
Invisible column.

There could be other usecases for Invisible in the future, perhaps
for the internal virtual columns used by hash-sharded indexes or
expression-based indexes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59233)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
> Reviewed 17 of 19 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59233)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/optbuilder/scope.go, line 748 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59233#-MRaXZ667ifrkP6lOnIb:-MRaXZ667ifrkP6lOnIc:b2kt9cu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec620e631a10b929f9bc50115ddfe13ee8f9a5a9/pkg/sql/opt/optbuilder/scope.go#L748)):*
<details><summary><i alt="summary">Quoted 27 lines of code…</i></summary>

> ```
> 			if col.visibility == cat.Invisible {
> 				// Act as if this column is not present so that matches in higher scopes
> 				// can be found. However, if no match is found in higher scopes and this
> 				// is a mutation column, report a backfill error rather than a "not
> 				// found" error.
> 				if col.mutation {
> 					reportBackfillError = true
> 				}
> 				continue
> 			}
> 			if col.table.ObjectName == "" && col.visibility == cat.Visible {
> 				if candidateFromAnonSource != nil {
> 					moreThanOneCandidateFromAnonSource = true
> 					break
> 				}
> 				candidateFromAnonSource = col
> 			} else if col.visibility == cat.Visible {
> 				if candidateWithPrefix != nil {
> 					moreThanOneCandidateWithPrefix = true
> 				}
> 				candidateWithPrefix = col
> 			} else if allowHidden {
> 				if hiddenCandidate != nil {
> 					moreThanOneHiddenCandidate = true
> 				}
> 				hiddenCandidate = col
> 			}
> ```
</details>

This code layout feels sort of suble/fragile in case some new visibility gets added. It's implicit that the `allowHidden` case is covering only the value of hidden because of the above continue

What do you think about reworking it into a switch 


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59233)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/optbuilder/scope.go, line 748 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59233#-MRaXZ667ifrkP6lOnIb:-MRaeUUH3AfieGoOb8yb:b-ohdo8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec620e631a10b929f9bc50115ddfe13ee8f9a5a9/pkg/sql/opt/optbuilder/scope.go#L748)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 			if col.visibility == cat.Invisible {
> 				// Act as if this column is not present so that matches in higher scopes
> 				// can be found. However, if no match is found in higher scopes and this
> 				// is a mutation column, report a backfill error rather than a "not
> 				// found" error.
> 				if col.mutation {
> 					reportBackfillError = true
> 				}
> 				continue
> 			}
> 			if col.table.ObjectName == "" && col.visibility == cat.Visible {
> 				if candidateFromAnonSource != nil {
> 					moreThanOneCandidateFromAnonSource = true
> 					break
> 				}
> 				candidateFromAnonSource = col
> 			} else if col.visibility == cat.Visible {
> 				if candidateWithPrefix != nil {
> 					moreThanOneCandidateWithPrefix = true
> 				}
> 				candidateWithPrefix = col
> 			} else if allowHidden {
> 				if hiddenCandidate != nil {
> 					moreThanOneHiddenCandidate = true
> 				}
> 				hiddenCandidate = col
> 			}
> ```
This code layout feels sort of suble/fragile in case some new visibility gets added. It's implicit that the `allowHidden` case is covering only the value of hidden because of the above continue

What do you think about reworking it into a switch 
</blockquote></details>

Done, good call.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
I renamed `Invisible` to `Inaccessible` at @andy-kimball's suggestion.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59233#-:-MRbDfz2Cnd9iFiWi1yh:by5lc4l)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @mgartner, and @rytaft)



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> Reviewed 12 of 19 files at r1, 8 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59233)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/cat/column.go, line 190 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59233#-MRbGkmU14C8KK2i2deK:-MRbGkmU14C8KK2i2deL:bcxlrec) ([raw file](https://github.com/cockroachdb/cockroach/blob/69e2a3cc1e5cd8a353e8ee5345c1da8e88ace5ce/pkg/sql/opt/cat/column.go#L190)):*
> ```Go
> // MaybeHidden is a helper constructor for either Visible or Hidden, depending
> // on a flag.
> func MaybeHidden(hidden bool) ColumnVisibility {
> ```

It's times like this where a ternary operator in Go would be nice. You could replace functions like this...
___
*[pkg/sql/opt/cat/utils.go, line 321 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59233#-MRbHsqO9hBWw09NY-Ys:-MRbHsqO9hBWw09NY-Yt:bsbaqw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/69e2a3cc1e5cd8a353e8ee5345c1da8e88ace5ce/pkg/sql/opt/cat/utils.go#L321)):*
> ```Go
> 	// Omit the visibility for mutation and virtual inverted columns, which are
> 	// always inacessible.
> 	if kind != WriteOnly && kind != DeleteOnly && kind != VirtualInverted {
> ```

Aren't system columns also inaccessible? Should this be simply `if kind == Ordinary { .. }`?
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 335 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59233#-MRbKoyF2NP2Oucu3zwO:-MRbKoyF2NP2Oucu3zwP:b-5z12mp) ([raw file](https://github.com/cockroachdb/cockroach/blob/69e2a3cc1e5cd8a353e8ee5345c1da8e88ace5ce/pkg/sql/opt/optbuilder/mutation_builder.go#L335)):*
> ```Go
> 			// If the primary key column is hidden, then we don't need to use it
> 			// for the distinct on.
> 			if col := primaryIndex.Column(i); col.Visibility() != cat.Hidden {
> ```

could a PK column be inaccessible? should this be `col.Visibility() == cat.Visible`?
___
*[pkg/sql/opt/optbuilder/testdata/select, line 1453 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59233#-MRbJo514YsaERnOWNeb:-MRbJo52B6BvpWd1TSai:bkepfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/69e2a3cc1e5cd8a353e8ee5345c1da8e88ace5ce/pkg/sql/opt/optbuilder/testdata/select#L1453)):*
So inaccessible columns can be written to? This is not super clear based on the comment for `cat.Inaccessible`, so maybe expand the comment a bit.

> // Inaccessible columns are not visible to queries in any way.

Would this update fail or succeed?

```
UPDATE inaccessible SET v = 1
```

Might be worth adding as a test case.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59233)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @mgartner, and @rytaft)

___
*[pkg/sql/opt/cat/column.go, line 190 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59233#-MRbGkmU14C8KK2i2deK:-MRbT7RuDqpbE4YrHaOb:brqvma1) ([raw file](https://github.com/cockroachdb/cockroach/blob/69e2a3cc1e5cd8a353e8ee5345c1da8e88ace5ce/pkg/sql/opt/cat/column.go#L190)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It's times like this where a ternary operator in Go would be nice. You could replace functions like this...
</blockquote></details>

Yeah, absolutely.. Waiting for those generics..
___
*[pkg/sql/opt/cat/utils.go, line 321 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59233#-MRbHsqO9hBWw09NY-Ys:-MRbR8Mz2A9osVYkA8Tz:b133unj) ([raw file](https://github.com/cockroachdb/cockroach/blob/69e2a3cc1e5cd8a353e8ee5345c1da8e88ace5ce/pkg/sql/opt/cat/utils.go#L321)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Aren't system columns also inaccessible? Should this be simply `if kind == Ordinary { .. }`?
</blockquote></details>

No, they are "hidden" (they need to be accessible, otherwise there's no point to them).
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 335 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59233#-MRbKoyF2NP2Oucu3zwO:-MRbRRgt3a7EAALhgsEa:bdnxq1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/69e2a3cc1e5cd8a353e8ee5345c1da8e88ace5ce/pkg/sql/opt/optbuilder/mutation_builder.go#L335)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

could a PK column be inaccessible? should this be `col.Visibility() == cat.Visible`?
</blockquote></details>

We don't have cases where a PK col is inaccessible. I can't say I understand this code very well, but I believe it infers that if a PK col is hidden, it must be "rowid" and then we don't need to run a distinct on. Added a TODO to investigate more.
___
*[pkg/sql/opt/optbuilder/scope.go, line 748 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59233#-MRaXZ667ifrkP6lOnIb:-MRbTpq-EcbLtN4jTJ0q:b-sc3x0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec620e631a10b929f9bc50115ddfe13ee8f9a5a9/pkg/sql/opt/optbuilder/scope.go#L748)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Done, good call.
</blockquote></details>

Oops I accidentally reverted that change, fixed.
___
*[pkg/sql/opt/optbuilder/testdata/select, line 1453 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59233#-MRbJo514YsaERnOWNeb:-MRbWvym0OYPIbKgaJd_:b-1kr5ih) ([raw file](https://github.com/cockroachdb/cockroach/blob/69e2a3cc1e5cd8a353e8ee5345c1da8e88ace5ce/pkg/sql/opt/optbuilder/testdata/select#L1453)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

So inaccessible columns can be written to? This is not super clear based on the comment for `cat.Inaccessible`, so maybe expand the comment a bit.

> // Inaccessible columns are not visible to queries in any way.

Would this update fail or succeed?

```
UPDATE inaccessible SET v = 1
```

Might be worth adding as a test case.
</blockquote></details>

Nice catch, should error out. Fixed and added more tests.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
Nice! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59233#-:-MRbe05jAfntM5KbVZY7:bfd61qr)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rytaft)



<!-- Sent from Reviewable.io -->

RaduBerinde committed:
opt: introduce inaccessible columns to opt catalog

There is a need to introduce a new column state during schema changes:
a state in which a column has been backfilled and is readable for
internal operations (like updating an index on that column during an
UPDATE), but is still not public.

Adding a new ColumnKind would complicate a lot of the optimizer code
and add more combinations that need testing. To avoid this, we note
that there is effectively no difference between such a column and an
ordinary column, other than it not being visible to queries.

This change extends the `IsHidden()` column property to a
`ColumnVisibility` which has three values (Visible, Hidden,
Inaccessible). The new column state will be presented as an Ordinary,
Inaccessible column.

There could be other usecases for Inaccessible in the future, perhaps
for the internal virtual columns used by hash-sharded indexes or
expression-based indexes.

Release note: None

RaduBerinde commented:
bors r+
